src/dex/verified/utils.ts

```diff
@@ -0,0 +1,172 @@
+  contractInterface: Interface,
+  functionName: string,
+  resultEntry: { success: boolean; returnData: any },
+  poolAddress: string,
+): Result {
+  if (!resultEntry.success)
+    throw new Error(`Failed to execute ${functionName} for ${poolAddress}`);
+  return contractInterface.decodeFunctionResult(
+    functionName,
+    resultEntry.returnData,
+  );
+}
+
+// export function poolGetPathForTokenInOut(
+//   tokenInAddress: string,
+//   tokenOutAddress: string,
+//   pool: SubgraphPoolBase,
+//   poolsMap: SubgraphPoolAddressDictionary,
+// ): BalancerPathHop[] {
+//   const tokenIn = findRequiredMainTokenInPool(tokenInAddress, pool);
+//   const tokenOut = findRequiredMainTokenInPool(tokenOutAddress, pool);
+
+//   const tokenInHops: BalancerPathHop[] = reverse([...tokenIn.pathToToken]).map(
+//     hop => ({
+//       pool: poolsMap[hop.poolAddress],
+//       tokenIn: hop.token,
+//       tokenOut: { address: hop.poolAddress, decimals: 18 },
+//     }),
+//   );
+
+//   const tokenOutHops: BalancerPathHop[] = tokenOut.pathToToken.map(hop => ({
+//     pool: poolsMap[hop.poolAddress],
+//     tokenIn: { address: hop.poolAddress, decimals: 18 },
+//     tokenOut: hop.token,
+//   }));
+
+//   return [
+//     ...tokenInHops,
+//     { pool, tokenIn: tokenIn.poolToken, tokenOut: tokenOut.poolToken },
+//     ...tokenOutHops,
+//   ];
+// }
+
+// export function getAllPoolsUsedInPaths(
+//   from: string,
+//   to: string,
+//   allowedPools: SubgraphPoolBase[],
+//   poolAddressMap: SubgraphPoolAddressDictionary,
+// ) {
+//   //get all pools from the nested paths
+//   return uniqBy(
+//     allowedPools
+//       .map(pool =>
+//         poolGetPathForTokenInOut(
+//           from.toLowerCase(),
+//           to.toLowerCase(),
+//           pool,
+//           poolAddressMap,
+//         ).map(hop => hop.pool),
+//       )
+//       .flat(),
+//     'address',
+//   );
+// }
+
+//Todo: confirm this handles both primary and secondary
+export function poolGetMainTokens(
+  pool: Omit<SubgraphPoolBase, 'mainTokens'>,
+  poolsMap: SubgraphPoolAddressDictionary,
+): SubgraphMainToken[] {
+  let mainTokens: SubgraphMainToken[] = [];
+
+  for (const token of pool.tokens) {
+    //skip
+    if (token.address === pool.address) {
+      continue;
+    }
+    const tokenPool = poolsMap[token.address];
+    if (tokenPool && isPrimaryPool(tokenPool.poolType)) {
+      //since primary main token is the security token used
+      const securityToken = pool.tokens.find(token => {
+        token.address === tokenPool.security;
+      })!;
+      mainTokens.push({
+        ...securityToken,
+        poolToken: token,
+        pathToToken: [
+          {
+            poolId: tokenPool.id,
+            poolAddress: tokenPool.address,
+            token: securityToken,
+          },
+        ],
+      });
+    } else if (tokenPool && isSecondaryPool(tokenPool.poolType)) {
+      //since secondary main token is the currency token used
```

Comment:
@Mohzcrea8me both primary and secondary pools can have security token as main token

Reply:
Oh okay. I will change it